doc/design/service-discovery.md

```diff
@@ -0,0 +1,421 @@
+    {
+        // Update local service registry
+        get();
+
+        return filterFor(serviceDescription);
+    }
+
+    // @todo check if still needed
+    bool offerService(const capro::ServiceDescription& serviceDescription) noexcept;
+    void stopOfferService(const capro::ServiceDescription& serviceDescription) noexcept;
+
+    ServiceRegistryTopic get() noexcept
+    {
+        m_subscriber.take().and_then([](auto& ServiceRegistryTopic){
+            // Update our local copy of the service registry
+            m_lastServiceRegistry = ServiceRegistryTopic;
+        }).or_else([](){
+            // No change in service registry
+        });
+        return m_lastServiceRegistry;
+    }
+
+    ServiceRegistryTopic getDelta() noexcept
+    {
+        auto old = m_lastServiceRegistry;
+        get();
+        return old - m_lastServiceRegistry;
+    }
+
+    ServiceRegistryTopic getDelta(const capro::ServiceDescription& serviceDescription) noexcept;
+
+   filterFor(const capro::ServiceDescription& serviceDescription) noexcept;
+  private:
+    Subscriber<ServiceRegistryTopic> m_subscriber{"ServiceDiscovery", "GlobalInstance", "ServiceRegistryTopic"};
+    ServiceRegistryTopic m_lastServiceRegistry;
+}
+
+// Class should be attachable to Listener
+iox::popo::DiscoveryInfo discoveryInfo;
+myListener.attachEvent(discoveryInfo, DATA_RECEIVED, createNotificationCallback(userDefinedCallback));
+```
+
+#### Event-based notification: Alternative E
+
+```cpp
+// StatusPort implementation
+StatusPortHeader {
+    using uint64_t INVALID {0};
+    using uint64_t UPDATING {-1};
+    std::atomic<uint64_t> generation{INVALID};
+}
+template <typename T>
+using StatusPortData = Sample<T, StatusPortHeader>; // pretty neat, eh? We use a chunk to store the StatusPortData -> this means the number of ports and data sizes can be configured by the user without recompiling RouDi; could potentially also be done for the other ports
+
+template <typename T>
+StatusPortWriter {
+  public:
+    void write(cxx::function_ref<void(T&)> f);
+  private:
+    StatusPortData<T> data; // potentially also just a `ChunkHeader*`
+};
+
+template <typename T>
+StatusPortReader {
+  public:
+    void read(cxx::function_ref<void(const T&)> f);
+  private:
+    StatusPortData<T> data;
+};
+
+// User facing API
+class DiscoveryInfo
+{
+  public:
+
+    // Move all service-related methods from PoshRuntime to this class
+    cxx::expected<InstanceContainer, FindServiceError>
+    PoshRuntime::findService(const capro::ServiceDescription& serviceDescription) noexcept
+    {
+        // Update local service registry
+        get();
+
+        return filterFor(serviceDescription);
+    }
+
+    // @todo check if still needed
+    bool offerService(const capro::ServiceDescription& serviceDescription) noexcept;
+    void stopOfferService(const capro::ServiceDescription& serviceDescription) noexcept;
+
+    ServiceRegistryTopic get() noexcept
```

Comment:
If the `statusPortReader` supports the callback based approach we should also make this available for the public interface since it could become handy.
Something like: `void getCallback(const cxx::function<const ServiceRegistryTopic&> f);` then the user can provide directly a free function which handles it. Something similar to `Publisher::publishResultOf`.

And `getCallback` could then be reused by all the `get` impls here.

Reply:
I'm unsure here. Usually I prefer to have one way only how the user can access an API and not several slightly different ones (Hyrums Law). I'll keep it in mind for the implementation.

Reply:
Here I would disagree since it would make the life of the user easier and we also did go exactly this way with the subscriber and the publisher.